src/Orleans/Logging/LoggerImpl.cs

```diff
@@ -0,0 +1,614 @@
+                              || (sev == Severity.Verbose || sev == Severity.Verbose2 || sev == Severity.Verbose3);
+
+            lock (this)
+            {
+                sinceInterval = now - lastBulkLogMessageFlush;
+                if (sinceInterval >= LogManager.BulkMessageInterval)
+                {
+                    // Take local copy of buffered log message counts, now that this bulk message compaction period has finished
+                    copyMessageCounts = recentLogMessageCounts;
+                    recentLogMessageCounts = new Dictionary<int, int>();
+                    lastBulkLogMessageFlush = now;
+                }
+
+                // Increment recent message counts, if appropriate
+                if (isExcluded)
+                {
+                    count = 1;
+                    // and don't track counts
+                }
+                else if (recentLogMessageCounts.ContainsKey(logCode))
+                {
+                    count = ++recentLogMessageCounts[logCode];
+                }
+                else
+                {
+                    recentLogMessageCounts.Add(logCode, 1);
+                    count = 1;
+                }
+            }
+
+            // Output any pending bulk compaction messages
+            if (copyMessageCounts != null && copyMessageCounts.Count > 0)
+            {
+                object[] args = new object[4];
+                args[3] = sinceInterval;
+
+                // Output summary counts for any pending bulk message occurrances
+                foreach (int ec in copyMessageCounts.Keys)
+                {
+                    int num = copyMessageCounts[ec] - LogManager.BulkMessageLimit;
+
+                    // Only output log codes which exceeded limit threshold
+                    if (num > 0)
+                    {
+                        args[0] = ec;
+                        args[1] = num;
+                        args[2] = (num == 1) ? "" : "s";
+
+                        WriteLogMessage(ec + LogManager.BulkMessageSummaryOffset, Severity.Info, "Log code {0} occurred {1} additional time{2} in the previous {3}", args, null);
+                    }
+                }
+            }
+
+            // Should the current log message be output?
+            return isExcluded || (count <= LogManager.BulkMessageLimit);
+        }
+
+        private static string FormatMessageText(string format, object[] args)
+        {
+            // avoids exceptions if format string contains braces in calls that were not
+            // designed to use format strings
+            return (args == null || args.Length == 0) ? format : String.Format(format, args);
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+        private void WriteLogMessage(int errorCode, Severity sev, string format, object[] args, Exception exception)
```

Comment:
This is hear because this logic requires logger state, which is per logger.  LogManager only deals with globally static logger data.  We could refactor this to be on the LogManager, taking the state of the logger, but the goal of this PR was to break the TraceLogger into separate classes, not refactor it's logic.  Moving this function would require refactoring that would introduce more risk with little, if any, advantage.